lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -431,6 +431,18 @@ resolveDeclRefExpr(UnresolvedDeclRefExpr *UDRE, DeclContext *DC) {
   auto Lookup = lookupUnqualified(DC, Name, Loc, LookupOptions);
 
   if (!Lookup) {
+    // Recover from not finding 'Self' in a nominal type, if we are
```

Comment:
I'm not sure it makes sense to do this as recovery rather than unilaterally. Recovery means that a top-level value named "Self" will take precedence over the language feature.